### ISSUE TITLE
Wrong result of dot function

### ISSUE THREAD

**Issue by giordano** [JuliaLang/julia#17869]
While working on the `LombScargle.jl` package, I noticed that in some cases the `dot` function gives a surprisingly imprecise result.  For example, the dot product between a vector and `ones` should be equal to the sum of elements of the vector, but this isn't the case:
```julia
julia> versioninfo()
Julia Version 0.6.0-dev.49
Commit bc4eba5 (2016-08-06 08:00 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)
s = sin(linspace(0.01, 10pi, 1001));
dot(ones(s), s)
# => 0.003408004595124931
sum(s)
# => 0.003408004595202869
sum(ones(s).*s) # This is slower than `dot`, but more precise
# => 0.003408004595202869
```

**Comment by yuyichao** [comment]
The difference is 6eps which sounds pretty good to me.

**Comment by giordano** [comment]
Not if you expect the result to pass a `@test_approx_eq` test (I know of `@test_approx_eq_eps`, but I don't know in advance the absolute eps)

**Comment by yuyichao** [comment]
> Not if you expect the result to pass a `@test_approx_eq` test.

I don't really think that's how the precision of these functions are determined and you need to come up with a better way to test your functions by for example manually determine the bound. The result is very sensitive to the order of `sum` and I don't think we can guarantee that all these functions are summing in exactly the same order.

**Comment by giordano** [comment]
A bit of background.  In order to ensure correctness of numerical result of the package, I'm comparing the output of `LombScargle.jl` with that of a similar Python library, but I had to skip a test (involving the use of `dot` in the Julia implementation) because the difference between of `LombScargle.jl` and the Python library is muuuch larger than just 6 eps (some 4 orders of magnitudes larger).  I suspected that the root of the problem could have been the `dot` function (because in any case I noticed this slightly imprecise result), but the outcome of `dot` function is simply multiplied by `0.5*N` (with `N` being the length of the vector, so an integer), thus now I don't expect that an error of ~10 eps is propagated into an error of ~10^5 eps with such a simple multiplication.

I'll try and find the culprit of this difference, and I leave you the decision whether to address the imprecise result of `dot`.
